util/reggen/gen_rust.py

```diff
@@ -0,0 +1,388 @@
+
+                statement = gen_const(dname, '_OFFSET', str(field.bits.lsb), existing_defines)
+                genout(outstr, statement)
+            if field.enum is not None:
+                for enum in field.enum:
+                    ename = as_define(enum.name)
+                    value = hex(enum.value)
+                    statement = gen_const(
+                        defname + '_' + as_define(field.name),
+                        '_VALUE_' + ename,
+                        value,
+                        existing_defines)
+                    genout(outstr, statement)
+    genout(outstr, '\n')
+    return
+
+
+def gen_const_window(outstr: TextIO,
+                     win: Window,
+                     comp: str,
+                     regwidth: int,
+                     rnames: Set[str],
+                     existing_defines: Set[str]) -> None:
+    offset = win.offset
+
+    genout(outstr, format_comment('Memory area: ' + first_line(win.desc)))
+    defname = as_define(comp + '_' + win.name)
+    genout(
+        outstr,
+        gen_const(defname, '_REG_OFFSET', hex(offset), existing_defines))
+    items = win.items
+    genout(
+        outstr,
+        gen_const(defname, '_SIZE_WORDS', str(items), existing_defines))
+    items = items * (regwidth // 8)
+    genout(
+        outstr,
+        gen_const(defname, '_SIZE_BYTES', str(items), existing_defines))
+
+    wid = win.validbits
+    if (wid != regwidth):
+        mask = (1 << wid) - 1
+        genout(outstr,
+               gen_const(defname, '_MASK ', hex(mask), existing_defines))
+
+
+def gen_rust_module_param(outstr: TextIO,
+                          param: LocalParam,
+                          module_name: str,
+                          existing_defines: Set[str]) -> None:
+    # Presently there is only one type (int), however if the new types are
+    # added, they potentially need to be handled differently.
+    known_types = ["int"]
+    if param.param_type not in known_types:
+        warnings.warn("Cannot generate a module define of type {}"
+                      .format(param.param_type))
+        return
+
+    if param.desc is not None:
+        genout(outstr, format_comment(first_line(param.desc)))
+    # Heuristic: if the name already has underscores, it's already snake_case,
+    # otherwise, assume StudlyCaps and covert it to snake_case.
+    param_name = param.name if '_' in param.name else to_snake_case(param.name)
+    define_name = as_define(module_name + '_PARAM_' + param_name)
+    if param.param_type == "int":
+        define = gen_const(define_name, '', param.value, existing_defines)
+
+    genout(outstr, define)
+    genout(outstr, '\n')
+
+
+def gen_const_module_params(outstr: TextIO,
+                            module_data: IpBlock,
+                            module_name: str,
+                            register_width: int,
+                            existing_defines: Set[str]) -> None:
+    module_params = module_data.params
```

Comment:
Nit: I don't think you really need this local: it's just used once, on the next line, and could equally well be inlined.

Reply:
Done.